Read version 51 (JDK 7) class files.

JamesIry opened the pull request:
This commit makes the ClassFileReader/ICodeReader parse class files
from JDK 7 (class file version 51). It does that by skipping over
the method handle related entries in the constant pool and by doing
some dummy processing on invoke dynamic instructions. The inliner
is updated to not try to inline a method with an invoke dynamic
instruction. A place holder INVOKE_DYNAMIC instruction is added to ICode
but it is designed to create an error if there's ever any attempt to
analyze it. Because the inliner is the only phase that ever tries
to analyze ICode instructions not generated from Scala source and
because Scala source will never emit an INVOKE_DYNAMIC, the place
holder INVOKE_DYNAMIC should never cause any errors.

A test is included that generates a class file with INVOKE_DYNAMIC
and then compiles Scala code that depends on it.

JamesIry committed:
Read version 51 (JDK 7) class files.

This commit makes the ClassFileReader/ICodeReader parse class files
from JDK 7 (class file version 51). It does that by skipping over
the method handle related entries in the constant pool and by doing
some dummy processing on invoke dynamic instructions. The inliner
is updated to not try to inline a method with an invoke dynamic
instruction. A place holder INVOKE_DYNAMIC instruction is added to ICode
but it is designed to create an error if there's ever any attempt to
analyze it. Because the inliner is the only phase that ever tries
to analyze ICode instructions not generated from Scala source and
because Scala source will never emit an INVOKE_DYNAMIC, the place
holder INVOKE_DYNAMIC should never cause any errors.

A test is included that generates a class file with INVOKE_DYNAMIC
and then compiles Scala code that depends on it.

JamesIry commented:
subsumes https://github.com/scala/scala/pull/2243, putting it into the 2.10.x branch

JamesIry commented:
review @adriaanm

paulp commented:
LGTM

paulp commented on src/compiler/scala/tools/nsc/symtab/classfile/ICodeReader.scala:
```diff
@@ -506,6 +506,12 @@ abstract class ICodeReader extends ClassfileParser {
+          containsInvokeDynamic = true
+          val poolEntry = in.nextChar
+          in.skip(2)
+          code.emit(INVOKE_DYNAMIC(poolEntry))
```
Can we get more in the habit of logging something when "placeholder" code paths are triggered?

paulp commented:
I'm pulling back my LGTM holding out for a call to log() where indicated, then it will LGTM again.

Commit:
Log when invokedynamic instruction is encountered

Based on the review of https://github.com/scala/scala/pull/2257, this
commit adds a debuglog when an invokedynamic instruction is found
during class file parsing as a reminder that the implementation is
just a place holder.

JamesIry commented:
Added a debuglog
